### DISCUSSION TITLE
Improve autodownsample and cliptoview performance

### DISCUSSION THREAD

**Discussion by Mriv31** [pyqtgraph/pyqtgraph#2710] [Ideas]
Hey !

I'm using pyqtgraph to display very large datasets and I find the autodownsample and cliptoview very useful. However, for larger datasets (>10⁷ pts) it is very slow and I found simple fixes which increase tremendously the performances, bypassing ```np.isfinite``` and ```np.searchsorted``` when they are not necessary. After those little fixes I very smoothly navigate in huge timeseries. 

This is how I increase a lot the performance in my fork : 

### About downsampling 

```
if self.opts['autoDownsample']:

            if containsNonfinite:  #MRIV31 EDIT : avoids recomputing isfinite if not necessary
                finite_x = x[np.isfinite(x)]  # ignore infinite and nan values
            else:
                finite_x = x

            if view_range is not None and len(finite_x) > 1:
                dx = float(finite_x[-1]-finite_x[0]) / (len(finite_x)-1)
                if dx != 0.0:
                    width = self.getViewBox().width()
                    if width != 0.0:  # autoDownsampleFactor _should_ be > 1.0
                        ds_float = max(1.0, abs(view_range.width() / dx / (width * self.opts['autoDownsampleFactor'])))
                        if math.isfinite(ds_float):
                            ds = int(ds_float)
                        if ds>3:
                            ds = int(np.exp(int(np.log(ds))))  # MRIV31 EDIT : only allows discrete values of ds spaced logarithmically and prevents being caught in a loop between viewrange updates and small ds variations.
```

### About ClipToView 

[EDIT : as @pijyoi noted  the poor performance of `np.searchsorted` was due to the fact that my xArray was of type `int`. Using `np.searchsorted` with a `int` array while the second argument is a `float` is extremely unefficient, hence my problem. So just make sure your xarray has a type `np.float64`. This should also be corrected in the code of pyqtgraph, replacing `np.arange(len(y))` by `np.arange(len(y),dtype=np.float64)` in the initialization of `PlotDataItem`when no x array is passed].
 
If the xArray is sorted and linear (proportional to indices), I do not use  ```np.searchsorted``` anymore. To do that I added a property  ```xArrayLinSorted``` to the class ``` PlotDataset```.  It can be set to ```True``` either by the user by passing the argument  ```xArrayLinSorted=True``` to the initlization of the ```PlotDataItem``` or is automatically be set in the case of the absence of xArray passed by the user (when initialization of a PlotDataItem directly use ```np.arange```) 

I also added the properties ```xOrigin``` and ```xIncrement``` which contains respectively ```x[0] ``` and ```x[1]-x[0]```

```
        if self.opts['clipToView']:
            if view is None or view.autoRangeEnabled()[0]:
                pass # no ViewBox to clip to, or view will autoscale to data range.
            else:

                if view_range is not None and len(x) > 1:

                    if (self._datasetMapped.xArrayLinSorted == True): #MRIV31 Edit
                        x0 = (view_range.left() - self._datasetMapped.xOrigin)/self._datasetMapped.xIncrement
                        x0 = int(x0) - ds
                        x0 = max(0,x0)

                        x1 = (view_range.right() - self._datasetMapped.xOrigin)/self._datasetMapped.xIncrement
                        x1 = int(x1) + ds

                    else:
                        x0 = np.searchsorted(x, view_range.left()) - ds
                        x0 = fn.clip_scalar(x0, 0, len(x)) # workaround
                        # x0 = np.clip(x0, 0, len(x))

                        x1 = np.searchsorted(x, view_range.right()) + ds
                        x1 = fn.clip_scalar(x1, x0, len(x))
                        # x1 = np.clip(x1, 0, len(x))
                    x = x[x0:x1]
                    y = y[x0:x1]
```

### PlotDataItem update problem bypass

Also there is a problem when you start a plot. The PlotDataItem gets updated before the autodownsample or downsample properties get passed to it by the ```PlotItem```. To avoid that I must repeat  the arguments ``` autoDownsample=1,downsample=None,downsampleMethod="subsample"``` in the ```plot``` function even if those properties are set in the ```PlotItem```.

### Not Fixed : many calls to ```_getDisplayDataset```
In general when you start a plot, ```_getDislayDataset``` is called around 5 times because of many functions calling updateItems. I do not think it is necessary and maybe you can find a workaround to increase performance.

### Further not tested ideas 

Downsambling with "subsample" does not take much time but for the others method you might want to avoid to repeat that operation at each update, for example by saving resampled arrays somewhere (in _datasetMapped ?) for later reuse

If the autosampling mode is off and the cliptoview mode is on, only compute the resampled array once and apply cliptoview on the resampled array. Do not resample because of cliptoview.

    add more restrictive conditions before recomputing all, like a parameter ("xViewRangeWasChangedALot") in here in the function _getDisplayDataset :

```
 if( self._datasetDisplay is not None and
            not (self.property('xViewRangeWasChanged') and self.opts['clipToView']) and
            not (self.property('xViewRangeWasChanged') and self.opts['autoDownsample']) and
            not (self.property('yViewRangeWasChanged') and self.opts['dynamicRangeLimit'] is not None)
        ):
```

Note that the condition above which forces the update has a redundancy. It it is also forced by the self._dataDisplay = None in viewRangeChanged

**Comment by pijyoi** [comment]
Does the xArrayLinSorted really help? ClipToView already assumes that x is sorted and a binary search of 1e7 elements would only take about 23 iterations?

The isfinite bypass seems like a straightforward improvement.

**Reply by Mriv31** [reply]
Yes, a lot. Here I compare with and without `xArrayLinSorted`,  navigating in a `PlotDataItem` with a dataset of 5e7 points, zooming in on a portion of 400 points, `autoDownsample=1` `clipToView = 1`. 

With ```xArrayLinSorted``` :
```
xArrayLinSorted is 1 
 _getDisplayDataset took 0.0003504753112792969s
```
Without ```xArrayLinSorted``` :
```
xArrayLinSorted is 0 
 _getDisplayDataset took 0.14357566833496094s
```

I  attach a screen capture of how fluid it now is to navigate is such a big dataset with those modifications :  

[Capture vidéo du 05-05-2023 14:39:06.webm](https://user-images.githubusercontent.com/24370506/236623603-2bd60d8a-017e-441b-859a-3b8237b68486.webm)


Note that I also added a method to PlotDataset() to inherit properties from another one and avoid to have to repeat it all the time :+1:     
``` 
def copyInheritedProperties(self,dataset):
        self.containsNonfinite = dataset.containsNonfinite
        self.xArrayLinSorted = dataset.xArrayLinSorted
        self.xIncrement = dataset.xIncrement
        self.xOrigin = dataset.xOrigin
```
and at the end of `getDataDisplay` 
```
self._datasetDisplay = PlotDataset( x,y )
self._datasetDisplay.copyInheritedProperties(self._dataset) #edit Mriv31
```

**Comment by pijyoi** [comment]
The poor performance seems to be because `x` is of type `int32` while the right argument is a `float`.
```python
In [9]: xi32 = np.arange(1e7, dtype=np.int32)
In [10]: xf32 = np.arange(1e7, dtype=np.float32)
In [11]: xf64 = np.arange(1e7, dtype=np.float64)
In [12]: %timeit np.searchsorted(xi32, 100.0)
33.4 ms ± 1.12 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
In [13]: %timeit np.searchsorted(xf32, 100.0)
32.7 ms ± 1.72 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
In [14]: %timeit np.searchsorted(xf64, 100.0)
2.57 µs ± 36.7 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

I suspect that if you were to replace
`x = np.arange(len(y))` with `x = np.arange(len(y), dtype=np.float64)`, you would get better performance.


Or pass in x manually with dtype float64.
```python
import pyqtgraph as pg
import numpy as np

pg.mkQApp()
pw = pg.PlotWidget()
pw.setClipToView(True)
pw.enableAutoRange(False)
pw.setDownsampling(ds=None, auto=True, mode="subsample")
data = np.arange(10000000)+np.random.normal(0,10,10000000)
x = np.arange(data.size, dtype=data.dtype)
pw.plot(x, data)
pw.show()
pg.exec()
```

**Reply by Mriv31** [reply]
Indeed !! Wonderful !!! But takes useless memory :-D
Note that you need to correct that in `PlotDataItem`  since the class initializes with `arange` when no x is passed as argument : 
```
    if y is None or len(y) == 0: # empty data is represented as None
            yData = None
        else: # actual data is represented by ndarray
            if not isinstance(y, np.ndarray):
                y = np.array(y)
            yData = y.view(np.ndarray)
            if x is None:
                x = np.arange(len(y)) //// HERE ADD dtype=np.float64
            
```

**Comment by pijyoi** [comment]
We could replace `np.searchsorted` with `bisect.bisect_left`.
This change would be transparent to the user.

```python
In [3]: xi32 = np.arange(1e7, dtype=np.int32)
In [4]: xf32 = np.arange(1e7, dtype=np.float32)
In [5]: xf64 = np.arange(1e7, dtype=np.float64)

In [7]: %timeit np.searchsorted(xi32, 100.0)
35.7 ms ± 5.28 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
In [8]: %timeit np.searchsorted(xf32, 100.0)
36.1 ms ± 5.49 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
In [9]: %timeit np.searchsorted(xf64, 100.0)
2.71 µs ± 123 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [10]: %timeit bisect.bisect_left(xi32, 100.0)
60 µs ± 4.93 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
In [11]: %timeit bisect.bisect_left(xf32, 100.0)
53.8 µs ± 1.76 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
In [12]: %timeit bisect.bisect_left(xf64, 100.0)
2.42 µs ± 114 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

**Comment by pijyoi** [comment]
I have created #2719 implementing 2 of the ideas in this discussion. Would appreciate it if you could test it out.

**Comment by pijyoi** [comment]
Regarding point 2 of your comment https://github.com/pyqtgraph/pyqtgraph/pull/2719#issuecomment-1544597880, would it not be sufficient for you to `setData` later?
```python
import pyqtgraph as pg
import numpy as np

pg.mkQApp()

pw = pg.PlotWidget()
pw.show()

pw.setClipToView(True)
pw.enableAutoRange(False)
pw.setDownsampling(ds=None, auto=True, mode="subsample")
data = np.arange(10000000)+np.random.normal(0,10,10000000)
pdi = pw.plot()
pdi.setData(data)

pg.exec()
```

A `print(ds, view_range)` gives
```
1 PySide6.QtCore.QRectF(0.000000, 0.000000, 1.000000, 1.000000)
3595 PySide6.QtCore.QRectF(-419914.670054, -48428.750128, 10839828.340107, 1096844.708175)
3632 PySide6.QtCore.QRectF(-421990.999841, -484054.804584, 10841676.999681, 10965783.625268)
3632 PySide6.QtCore.QRectF(-422041.692644, -484113.296609, 10842979.385289, 10967108.739358)
```

where it was previously
```
1 None
1 PySide6.QtCore.QRectF(0.000000, 0.000000, 1.000000, 1.000000)
1 PySide6.QtCore.QRectF(0.000000, 0.000000, 1.000000, 1.000000)
1 PySide6.QtCore.QRectF(-419914.670054, -48426.812238, 10839828.340107, 1096842.680794)
3595 PySide6.QtCore.QRectF(-419914.670054, -48426.812238, 10839828.340107, 1096842.680794)
3595 PySide6.QtCore.QRectF(-419914.670054, -48426.812238, 10839828.340107, 1096842.680794)
3594 PySide6.QtCore.QRectF(-419817.921704, -484052.378436, 10837330.843408, 10965770.536599)
3607 PySide6.QtCore.QRectF(-420572.780673, -484092.450966, 10839653.561346, 10966678.364187)
3607 PySide6.QtCore.QRectF(-420576.818774, -484097.099809, 10839757.637548, 10966783.681926)
```

**Reply by Mriv31** [reply]
Yes it is the perfect solution. Also solves point number 1. 
We should make it an example "large datasets". Sooooooo nice to be able to load 10^8 pts in a few seconds and then browse into them extremely fast.
Thanks @pijyoi.
Still remains the infinite loop sometimes (had it on windows too). The  `round ` that you suggested does not solve the problem, just shifts it to other combination of view sizes and point numbers.

 ` round(number/10)*10` is not great  for small ds.

Could do the closest power of two:   `ds = int(2**(int(np.log2(ds))))`, which seems an intuitive solution. 

But again with all those solutions it is possible that for the right parameter it might oscillate (but much less often than when all integers are allowed). 

We could also just break the loop. I think the loop is due to the fact that the max values of the displayed dataset change a little with ds. One solution would be to always add the last point to the downsampled dataset, however I think this is no more basic slicing, and will thus create a copy instead of a view.

**Comment by lidstrom83** [comment]
Hi everyone. Some of this discussion seem to resemble what I was trying in PR #1574. I thought I'd mention it in case you haven't seen it. The approach there seemed to work pretty well, but I never got around to finishing the PR.

**Reply by Mriv31** [reply]
Hi Lidstrom. 

Indeed very similar question !

Note that PR [2719](https://github.com/pyqtgraph/pyqtgraph/pull/2719) by @pijyoi  + its hint above (first create the plot, then `setData`), already makes a huge difference.

Try to pull the last dev version, load 10⁸ or 10⁹ points dataset this way and navigate using pan and zoom and you will see. Now very fluid.

**Comment by pijyoi** [comment]
Something like the following would break the oscillation:
```python
        if not hasattr(self, '_last_ds_value'):
            self._last_ds_value = ds
        if abs(self._last_ds_value - ds) / ds < 0.05:
            ds = self._last_ds_value
        self._last_ds_value = ds
        print(ds, view_range)
```
The stored state would of course need to be reset at appropriate moments.
